core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -294,13 +294,25 @@ private void writeVersionHint(int versionToWrite) {
     Path versionHintFile = versionHintFile();
     FileSystem fs = getFileSystem(versionHintFile, conf);
 
-    try (FSDataOutputStream out = fs.create(versionHintFile, true /* overwrite */)) {
-      out.write(String.valueOf(versionToWrite).getBytes(StandardCharsets.UTF_8));
+    try {
+      if (fs.exists(versionHintFile)) {
+        Path tempVersionHintFile = metadataPath("version-hint.temp" );
+        writeFileToPath(fs, tempVersionHintFile);
+        fs.delete(versionHintFile, false);
+        fs.rename(tempVersionHintFile, versionHintFile);
+      } else {
+        writeFileToPath(fs, versionHintFile);
+      }
     } catch (IOException e) {
       LOG.warn("Failed to update version hint", e);
     }
   }
 
+  private void writeFileToPath(FileSystem fs, Path path) throws IOException{
```

Comment:
Formatting checks might fail here.

I have learned that it is worth to run checkstyle checks before committing:
```
./gradlew build -x test
```